File: packages/vscode-extension/src/runtime-environment.js
"use strict";

/**
 * Installation Doctor environment probes. Process and filesystem access stay
 * here so the report renderer and compatibility classifier remain pure.
 */

const fs = require("node:fs");
const os = require("node:os");
const path = require("node:path");
const { spawn } = require("node:child_process");
const { MIN_NODE_VERSION, compareVersions } = require("./version-check");

function parseNodeVersion(output) {
  const match = /\bv?(\d+\.\d+\.\d+)\b/.exec(String(output || ""));
  return match ? match[1] : null;
}

function parseJavaVersion(output) {
  const text = String(output || "");
  const quoted = /\bversion\s+"([^"]+)"/i.exec(text);
  if (quoted) return quoted[1].slice(0, 64);
  const open = /\b(?:openjdk|java)\s+([0-9][^\s]*)/i.exec(text);
  return open ? open[1].slice(0, 64) : null;
}

function runVersion(command, args, deps = {}) {
  const spawnImpl = deps.spawn || spawn;
  const timeoutMs = Number(deps.timeoutMs) > 0 ? Number(deps.timeoutMs) : 5000;
  return new Promise((resolve) => {
    let done = false;
    let output = "";
    const finish = (value) => {
      if (done) return;
      done = true;
      resolve(value);
    };
    try {
      const child = spawnImpl(command, args, {
        windowsHide: true,
        shell: false,
        env: deps.env || process.env,
      });
      const collect = (chunk) => {
        output += String(chunk || "");
        if (output.length > 8192) output = output.slice(0, 8192);
      };
      child.stdout?.on("data", collect);
      child.stderr?.on("data", collect);
      child.on("error", () => finish(null));
      child.on("close", (code) =>
        finish(code === 0 ? output.trim() || null : null),
      );
      const timer = setTimeout(() => {
        try {
          child.kill();
        } catch {
          // best effort
        }
        finish(null);
      }, timeoutMs);
      timer.unref?.();
    } catch {
      finish(null);
    }
  });
}

function defaultPluginRegistryCacheDir({
  platform = process.platform,
  homeDir = os.homedir(),
  env = process.env,
} = {}) {
  const app = "chainlesschain-desktop-vue";
  if (platform === "win32") {
    return path.join(
      env.APPDATA || path.join(homeDir, "AppData", "Roaming"),
      app,
      "plugin-registry-cache",
    );
  }
  if (platform === "darwin") {
    return path.join(
      homeDir,
      "Library",
      "Application Support",
      app,
      "plugin-registry-cache",
    );
  }
  return path.join(
    env.XDG_CONFIG_HOME || path.join(homeDir, ".config"),
    app,
    "plugin-registry-cache",
  );
}

function countRegularFiles(dir, fsImpl = fs) {
  try {
    return fsImpl
      .readdirSync(dir, { withFileTypes: true })
      .filter((entry) => entry?.isFile?.()).length;
  } catch {
    return 0;
  }
}

function inspectManagedCli(rootDir, fsImpl = fs) {
  if (!rootDir) return { status: "unconfigured", version: null };
  const statePath = path.join(rootDir, "current.json");
  if (!fsImpl.existsSync(statePath)) {
    return { status: "missing", version: null };
  }
  try {
    const state = JSON.parse(fsImpl.readFileSync(statePath, "utf8"));
    const version =
      typeof state?.version === "string" ? state.version.slice(0, 64) : null;
    if (!version) return { status: "corrupt", version: null };
    const packageDir = path.join(rootDir, version, "package");
    return {
      status: fsImpl.existsSync(packageDir) ? "ready" : "incomplete",
      version,
      rollbackVersion:
        typeof state.previousVersion === "string"
          ? state.previousVersion.slice(0, 64)
          : null,
    };
  } catch {
    return { status: "corrupt", version: null };
  }
}

function inspectOfflineCaches({
  managedCliRoot,
  pluginRegistryCacheDir = defaultPluginRegistryCacheDir(),
  fsImpl = fs,
} = {}) {
  const registryEntries = countRegularFiles(pluginRegistryCacheDir, fsImpl);
  return {
    managedCli: inspectManagedCli(managedCliRoot, fsImpl),
    pluginRegistry: {
      status: registryEntries > 0 ? "ready" : "missing",
      entries: registryEntries,
    },
  };
}

async function probeRuntimeEnvironment(options = {}) {
  const run =
    typeof options.runVersion === "function"
      ? options.runVersion
      : (command, args) => runVersion(command, args, options);
  const [nodeOutput, javaOutput] = await Promise.all([
    run("node", ["--version"]),
    run("java", ["-version"]),
  ]);
  const nodeVersion = parseNodeVersion(nodeOutput);
  const javaVersion = parseJavaVersion(javaOutput);
  let nodeStatus = "missing";
  if (nodeVersion) {
    nodeStatus =
      compareVersions(nodeVersion, MIN_NODE_VERSION) >= 0
        ? "ready"
        : "outdated";
  }
  return {
    node: {
      status: nodeStatus,
      version: nodeVersion,
      minimumVersion: MIN_NODE_VERSION,
    },
    java: {
      status: javaVersion ? "ready" : "missing",
      version: javaVersion,
    },
    caches: inspectOfflineCaches(options),
  };
}

function formatRuntimeEnvironment(environment = {}) {
  const node = environment.node || {};
  const java = environment.java || {};
  const caches = environment.caches || {};
  const managed = caches.managedCli || {};
  const registry = caches.pluginRegistry || {};
  return [
    `- Node.js: ${node.version || "missing"}${
      node.minimumVersion ? ` (minimum ${node.minimumVersion})` : ""
    } — ${node.status || "unknown"}`,
    `- Java on PATH: ${java.version || "missing"} — ${
      java.status || "unknown"
    }`,
    `- Managed CLI offline copy: ${managed.status || "unknown"}${
      managed.version ? ` (${managed.version})` : ""
    }${managed.rollbackVersion ? `; rollback ${managed.rollbackVersion}` : ""}`,
    `- Plugin registry offline cache: ${registry.status || "unknown"}${
      Number.isFinite(Number(registry.entries))
        ? ` (${Number(registry.entries)} entries)`
        : ""
    }`,
  ];
}

module.exports = {
  parseNodeVersion,
  parseJavaVersion,
  runVersion,
  defaultPluginRegistryCacheDir,
  inspectManagedCli,
  inspectOfflineCaches,
  probeRuntimeEnvironment,
  formatRuntimeEnvironment,
};

File: packages/jetbrains-plugin/src/main/java/com/chainlesschain/ide/RuntimeEnvironment.java
package com.chainlesschain.ide;

import java.io.BufferedReader;
import java.io.File;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/** Node/Java probes and offline-cache diagnosis for Installation Doctor. */
public final class RuntimeEnvironment {
    private static final Pattern VERSION =
            Pattern.compile("\\bv?(\\d+\\.\\d+\\.\\d+)\\b");
    private static final Pattern JAVA_QUOTED =
            Pattern.compile("\\bversion\\s+\"([^\"]+)\"", Pattern.CASE_INSENSITIVE);
    private static final Pattern JAVA_OPEN =
            Pattern.compile("\\b(?:openjdk|java)\\s+([0-9][^\\s]*)",
                    Pattern.CASE_INSENSITIVE);

    private RuntimeEnvironment() {}

    public static final class Result {
        public final String nodeStatus;
        public final String nodeVersion;
        public final String minimumNodeVersion;
        public final String javaStatus;
        public final String javaVersion;
        public final String ideJavaVersion;
        public final String managedCliStatus;
        public final String managedCliVersion;
        public final String rollbackVersion;
        public final String pluginRegistryStatus;
        public final int pluginRegistryEntries;

        Result(String nodeStatus, String nodeVersion,
                String minimumNodeVersion, String javaStatus,
                String javaVersion, String ideJavaVersion,
                String managedCliStatus, String managedCliVersion,
                String rollbackVersion, String pluginRegistryStatus,
                int pluginRegistryEntries) {
            this.nodeStatus = nodeStatus;
            this.nodeVersion = nodeVersion;
            this.minimumNodeVersion = minimumNodeVersion;
            this.javaStatus = javaStatus;
            this.javaVersion = javaVersion;
            this.ideJavaVersion = ideJavaVersion;
            this.managedCliStatus = managedCliStatus;
            this.managedCliVersion = managedCliVersion;
            this.rollbackVersion = rollbackVersion;
            this.pluginRegistryStatus = pluginRegistryStatus;
            this.pluginRegistryEntries = pluginRegistryEntries;
        }
    }

    public static Result probe(File cwd) {
        String node = runVersion(Arrays.asList("node", "--version"), cwd, 5000);
        String java = runVersion(Arrays.asList("java", "-version"), cwd, 5000);
        return evaluate(
                node,
                java,
                System.getProperty("java.version"),
                ManagedCliRuntime.defaultRootDir(),
                defaultPluginRegistryCacheDir());
    }

    static Result evaluate(String nodeOutput, String javaOutput,
            String ideJavaVersion, String managedRoot,
            String registryCacheDir) {
        String nodeVersion = parseNodeVersion(nodeOutput);
        String nodeStatus = nodeVersion == null
                ? "missing"
                : CliVersionCheck.compare(
                        nodeVersion, ManagedCliRuntime.MIN_NODE_VERSION) >= 0
                                ? "ready" : "outdated";
        String javaVersion = parseJavaVersion(javaOutput);

        String managedStatus = "unconfigured";
        String managedVersion = null;
        String rollbackVersion = null;
        if (managedRoot != null && !managedRoot.isEmpty()) {
            Path statePath = Paths.get(managedRoot, ManagedCli.STATE_FILE);
            if (!Files.isRegularFile(statePath)) {
                managedStatus = "missing";
            } else {
                try {
                    Object parsed = MiniJson.parse(
                            Files.readString(statePath, StandardCharsets.UTF_8));
                    if (!(parsed instanceof Map)) {
                        managedStatus = "corrupt";
                    } else {
                        Map<?, ?> state = (Map<?, ?>) parsed;
                        managedVersion = string(state.get("version"));
                        rollbackVersion = string(state.get("previousVersion"));
                        if (managedVersion == null) {
                            managedStatus = "corrupt";
                        } else {
                            Path packageDir = Paths.get(
                                    managedRoot, managedVersion, "package");
                            managedStatus = Files.isDirectory(packageDir)
                                    ? "ready" : "incomplete";
                        }
                    }
                } catch (Throwable invalid) {
                    managedStatus = "corrupt";
                }
            }
        }

        int registryEntries = countRegularFiles(registryCacheDir);
        return new Result(
                nodeStatus,
                nodeVersion,
                ManagedCliRuntime.MIN_NODE_VERSION,
                javaVersion == null ? "missing" : "ready",
                javaVersion,
                truncate(ideJavaVersion, 64),
                managedStatus,
                managedVersion,
                rollbackVersion,
                registryEntries > 0 ? "ready" : "missing",
                registryEntries);
    }

    public static List<String> formatLines(Result result) {
        Result r = result != null
                ? result
                : new Result(
                        "unknown", null, ManagedCliRuntime.MIN_NODE_VERSION,
                        "unknown", null, null, "unknown", null, null,
                        "unknown", 0);
        List<String> out = new ArrayList<>();
        out.add("Node.js: " + value(r.nodeVersion)
                + " (minimum " + r.minimumNodeVersion + ") — "
                + r.nodeStatus);
        out.add("Java on PATH: " + value(r.javaVersion)
                + " — " + r.javaStatus);
        out.add("IDE Java runtime: " + value(r.ideJavaVersion));
        out.add("Managed CLI offline copy: " + r.managedCliStatus
                + (r.managedCliVersion == null
                        ? "" : " (" + r.managedCliVersion + ")")
                + (r.rollbackVersion == null
                        ? "" : "; rollback " + r.rollbackVersion));
        out.add("Plugin registry offline cache: " + r.pluginRegistryStatus
                + " (" + r.pluginRegistryEntries + " entries)");
        return out;
    }

    static String parseNodeVersion(String output) {
        Matcher match = VERSION.matcher(output == null ? "" : output);
        return match.find() ? match.group(1) : null;
    }

    static String parseJavaVersion(String output) {
        String text = output == null ? "" : output;
        Matcher quoted = JAVA_QUOTED.matcher(text);
        if (quoted.find()) return truncate(quoted.group(1), 64);
        Matcher open = JAVA_OPEN.matcher(text);
        return open.find() ? truncate(open.group(1), 64) : null;
    }

    private static String runVersion(
            List<String> command, File cwd, long timeoutMs) {
        try {
            ProcessBuilder builder = new ProcessBuilder(command);
            if (cwd != null) builder.directory(cwd);
            CliLauncher.augmentPath(builder);
            builder.redirectErrorStream(true);
            Process process = builder.start();
            StringBuffer output = new StringBuffer();
            Thread pump = new Thread(() -> {
                try (BufferedReader reader = new BufferedReader(
                        new InputStreamReader(
                                process.getInputStream(),
                                StandardCharsets.UTF_8))) {
                    String line;
                    while ((line = reader.readLine()) != null
                            && output.length() < 8192) {
                        output.append(line).append('\n');
                    }
                } catch (Exception ignored) {
                    // best effort
                }
            }, "cc-runtime-environment-probe");
            pump.setDaemon(true);
            pump.start();
            boolean done = process.waitFor(timeoutMs, TimeUnit.MILLISECONDS);
            if (!done) {
                process.destroyForcibly();
                return "";
            }
            pump.join(500);
            return process.exitValue() == 0 ? output.toString() : "";
        } catch (Exception unavailable) {
            return "";
        }
    }

    private static int countRegularFiles(String dir) {
        if (dir == null || dir.isEmpty()) return 0;
        int count = 0;
        try (DirectoryStream<Path> stream =
                     Files.newDirectoryStream(Paths.get(dir))) {
            for (Path entry : stream) {
                if (Files.isRegularFile(entry)) count++;
            }
        } catch (Exception unavailable) {
            return 0;
        }
        return count;
    }

    private static String defaultPluginRegistryCacheDir() {
        String home = System.getProperty("user.home", ".");
        String app = "chainlesschain-desktop-vue";
        if (File.separatorChar == '\\') {
            String appData = System.getenv("APPDATA");
            if (appData == null || appData.isEmpty()) {
                appData = Paths.get(home, "AppData", "Roaming").toString();
            }
            return Paths.get(appData, app, "plugin-registry-cache").toString();
        }
        if (System.getProperty("os.name", "")
                .toLowerCase().contains("mac")) {
            return Paths.get(home, "Library", "Application Support",
                    app, "plugin-registry-cache").toString();
        }
        String xdg = System.getenv("XDG_CONFIG_HOME");
        if (xdg == null || xdg.isEmpty()) {
            xdg = Paths.get(home, ".config").toString();
        }
        return Paths.get(xdg, app, "plugin-registry-cache").toString();
    }

    private static String string(Object value) {
        return value instanceof String
                && !((String) value).trim().isEmpty()
                        ? truncate((String) value, 64) : null;
    }

    private static String value(String value) {
        return value == null || value.isEmpty() ? "missing" : value;
    }

    private static String truncate(String value, int max) {
        if (value == null) return null;
        return value.length() <= max ? value : value.substring(0, max);
    }
}
